### ISSUE TITLE
Product Button: Mini-Cart focus shift interrupts add-to-cart announcements

### ISSUE THREAD

**Issue by Aljullu** [woothemes/woocommerce#62228]
Issue found by @samueljseay in https://github.com/woocommerce/woocommerce/pull/62014#pullrequestreview-3480916551.

When the mini cart is configured to open automatically after adding a product, the focus shift to the mini cart drawer interrupts the screen-reader announcement from the Product Button which indicates that the product was added to the cart.

#### Write steps here

1. Go to **Appearance** > **Editor** > **Templates** > edit any template and select the Mini-Cart block.
2. Enable the **Behavior: Open drawer when adding** attribute.
3. On the frontend, go to the *Shop* page, add a product to the cart while using a screen reader.
4. Observe that the mini cart opens and immediately takes focus.
5. Notice that the screen-reader announcement is interrupted or not read at all.

#### A clear and concise description of what you expected to happen.

The announcement that a product has been added to the cart should not be interrupted. The focus shift caused by the mini cart opening should not cancel or override the message.

**Comment by Aljullu** [comment]
What's happening is that:

1. When the item is added to cart, we trigger `speak( 'Added to cart' )` to announce it to screen readers.
2. In parallel, the _Mini-Cart_ drawer is opened and receives focus, which causes screen readers to interrupt any previous announcements and to start announcing the drawer contents.

We could fix it by:

1. Delaying the drawer opening so we give enough time for 'Added to cart' to be announced.
2. Triggering `speak( 'Added to cart' )` from inside the _Mini-Cart_ drawer.

However, I wanted to step back and actually challenge _what_ we are trying to fix. If after adding to cart the _Mini-Cart_ is opened announcing with `Your cart (items: 123)`, is there any benefit of also announcing `Added to cart`? It feels redundant to me.

In the same way that if adding to cart redirects to the cart page, we aren't announcing that the product was added to cart, because that information will already be in the cart, I don't think we should do so with the _Mini-Cart_.

cc @samueljseay what do you think?

**Comment by Aljullu** [comment]
I will go ahead and close this issue based on the previous comment.

**Comment by samueljseay** [comment]
@Aljullu Sorry I didn't comment on this earlier. I was formulating a response and this dropped off my radar. I only saw it again due to the closing

From an accessibility perspective I think this is incorrect though: 

> In the same way that if adding to cart redirects to the cart page, we aren't announcing that the product was added to cart, because that information will already be in the cart, I don't think we should do so with the Mini-Cart.

The WCAG and accessibility best practices recommend that you don’t interrupt or override live announcements in assistive tech contexts without clear reason and this has no clear reason for happening, and the announcement should not be broken in my opinion.

A good compromise could be not to prevent the mini-cart from opening, but defer focus transfer until after the screen-reader announcement concludes. The focus transfer is mostly for assistive purposes anyway so this would treat assistive users the best we can.

**Comment by Aljullu** [comment]
Thanks for the feedback, @samueljseay!

> The WCAG and accessibility best practices recommend that you don’t interrupt or override live announcements in assistive tech contexts without clear reason and this has no clear reason for happening, and the announcement should not be broken in my opinion.

My point is that the live announcement is unnecessary when the Mini-Cart is opened and focused on right after adding to cart because it's clear that the product was added to cart.

On the other hand, I agree the '*Added to cart*' announcement is necessary if nothing happens after adding to cart (and that's why I created [https://github.com/woocommerce/woocommerce/issues/59438](<https://github.com/woocommerce/woocommerce/issues/59438>) in the first place 😄). But the fact that we are opening the *Mini-Cart* already signals that the action succeeded. Basically, instead of announcing "*Added to cart*" we are announcing "*Your cart (1 item)*", which is even more useful, IMO.

> A good compromise could be not to prevent the mini-cart from opening, but defer focus transfer until after the screen-reader announcement concludes. The focus transfer is mostly for assistive purposes anyway so this would treat assistive users the best we can.

I'm a bit wary of having the drawer open but not having the focus, even if it's just for a few seconds. I fear that brings other problems, like breaking Tab navigation. I'm also not sure how we would know how long to wait for the focus to be moved. Different locales and different screen readers might take a different amount of time to read the announcement, no?

That would be solved by triggering `speak( 'Added to cart' )` from inside the *Mini-Cart* drawer, once it's already open and focused. But then, that's when I wonder if it's really needed: if I click the Add to Cart button and the next thing that is announced to me is '*Your cart (1 item*)', isn't it obvious that the product was added to the cart? Does announcing '*Added to cart*' add any valuable information? 🤔

I think we should be mindful that making more announcements isn't always better if they don't add any value. What do you think?

**Comment by samueljseay** [comment]
> On the other hand, I agree the 'Added to cart' announcement is necessary if nothing happens after adding to cart (and that's why I created https://github.com/woocommerce/woocommerce/issues/59438 in the first place 😄). But the fact that we are opening the Mini-Cart already signals that the action succeeded. Basically, instead of announcing "Added to cart" we are announcing "Your cart (1 item)", which is even more useful, IMO.

I understand your point but I still disagree from accessibility perspective that opening Mini Cart implies anything happened successfully to a user using assistive technologies. It would at minimum be confusing.

> That would be solved by triggering speak( 'Added to cart' ) from inside the Mini-Cart drawer, once it's already open and focused. But then, that's when I wonder if it's really needed: if I click the Add to Cart button and the next thing that is announced to me is 'Your cart (1 item)', isn't it obvious that the product was added to the cart? Does announcing 'Added to cart' add any valuable information? 🤔

That only works for the I had nothing in cart I went to 1 item in cart. If they had multiple items in cart announcing how many items are in cart doesn't imply the increase unless they somehow memorized it?

> I think we should be mindful that making more announcements isn't always better if they don't add any value. What do you think?

I think they totally add value here.

> I'm a bit wary of having the drawer open but not having the focus, even if it's just for a few seconds. I fear that brings other problems, like breaking Tab navigation. I'm also not sure how we would know how long to wait for the focus to be moved. Different locales and different screen readers might take a different amount of time to read the announcement, no?

Not sure I agree. From WCAG perspective better not to interrupt. Tab navigation gives the most help to users with assistive tech, I'm sure they expect an announcement to complete due to the way they expect announcements not to be interrupted and how focus does interrupt that? That said I'm not 100% one way on that, we could announce from Mini Cart and it'd still be better than nothing.

**Comment by Aljullu** [comment]
> I understand your point but I still disagree from accessibility perspective that opening Mini Cart implies anything happened successfully to a user using assistive technologies. It would at minimum be confusing.

Hmmm, I'm still not fully convinced. If we don't consider opening the Mini-Cart as a way to signal "add-to-cart success", we have the same issue for users not using a screen reader, don't we?

For example, in this interaction, what kind of signal does the shopper get that another *Polo* was added to cart?

[Enregistrament de pantalla a 2026-01-26 12-37-32.mp4](https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/48de5788-afa0-43d2-80ac-82abb61c8709/e84df79f-99e0-4f8b-8267-40e1bb449aba?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi80OGRlNTc4OC1hZmEwLTQzZDItODBhYy04MmFiYjYxYzg3MDkvZTg0ZGY3OWYtOTllMC00ZjhiLTgyNjctNDBlMWJiNDQ5YWJhIiwiaWF0IjoxNzY5NDQ5NjUzLCJleHAiOjE4MDEwMjAyMTN9.GMzCiMdz9YcUNUDpkxirS_PK0KHfSmNboOFgGlF6rj4)

The shopper needs to scroll through the Mini-Cart to notice the quantity changed, the same as a user using a screen reader.

I think it's a common pattern that redirecting to Cart or opening the Mini-Cart after adding to cart is considered a way to signal success. But if we don't consider that, I don't think that issue is specific to screen reader users.

As a side note, there were some design proposals some time ago to change the add-to-cart behaviors by adding notices or changing the Mini-Cart title to *Added to cart*. There is this issue [https://github.com/woocommerce/woocommerce/issues/57194](<https://github.com/woocommerce/woocommerce/issues/57194>) (with more context in the links in [WOOPLUG-3896](https://linear.app/a8c/issue/WOOPLUG-3896/implement-a-confirmation-banner-so-it-displays-when-products-are-added)). Both approaches would keep the experience consistent between screen reader users and non-screen reader users, which I think is preferable.

Curious to know your thoughts, but my thinking is that it's not worth working on this issue (assuming it's an issue, which I don't fully agree with yet 😅) considering there are proposals to redesign the add-to-cart behavior that would redesign how adding to cart is announced to shoppers using and not using screen readers. What do you think?

**Comment by samueljseay** [comment]
Just prefixing this by saying, I'm not certain how big a deal this is, I'm advocating for visually impaired users here and while I might argue with you I find it hard to say how critical this is if you tried to rate the severity. I believe for visually impaired users this is quite important, (even just consider WCAG guidelines) and I will try argue for that, but I don't want to come across pushy either. :) 

> Hmmm, I'm still not fully convinced. If we don't consider opening the Mini-Cart as a way to signal "add-to-cart success", we have the same issue for users not using a screen reader, don't we?

I'm not sure I understand, this is not related to non screen reader users. I think (imho) that you're conflating the needs of non-impaired users with impaired users. As a sort of analogy, this is like a wheel chair ramp approach, we don't take away the ramp because the user who can walk doesn't need it. So I am just not sure your argument is in the right direction here.

> For example, in this interaction, what kind of signal does the shopper get that another Polo was added to cart?

A user with vision doesn't need the signal since they can see everything all at once, including the increment of the number in the button itself, which is (I believe) essentially what the announcement is compensating for. The whole UI guides them via site and visual indicators and such that they can process immediately.

Opening mini cart can be the result of clicking the mini cart icon, which indicates nothing to a visually impaired user. Now I can see your point, that it *implies* something happened in this case, but for a visually impaired user I would rather tell them what and how many were added to cart I think this is a really helpful cue for them. They cannot easily browse the interface at glance like a sighted user. These users have little isolated glances at whats happening in the UI and can only rely on these audio cues and so the clearer they are the better.

> Curious to know your thoughts, but my thinking is that it's not worth working on this issue (assuming it's an issue, which I don't fully agree with yet 😅) considering there are proposals to redesign the add-to-cart behavior that would redesign how adding to cart is announced to shoppers using and not using screen readers. What do you think?

I am really trying to be objective about this one and no hard feelings, I could wave it away and say no big deal, but if we are serious about assistive technologies, I think this matters as much as any other WCAG related thing we work on.

> Both approaches would keep the experience consistent between screen reader users and non-screen reader users, which I think is preferable.

Yep that could work, it's just my feeling is there should be an issue for this (or this should be captured in an issue to not lose it), even if this is reimagined in any refactor of the cart with additional notices. As long as there is an uninterrupted audio cue for the visually impaired user, I think thats good.

**Comment by Aljullu** [comment]
> A user with vision doesn't need the signal since they can see everything all at once, including the increment of the number in the button itself, which is (I believe) essentially what the announcement is compensating for.

Well, my point was that, as a sighted user:

1. I can't see the button increment because the Mini-Cart covers the button before the new quantity is visible.
2. The product that was recently added isn't visible in the list of products because it's below the fold.

So if we consider opening the Mini-Cart is not a clear indicator that the product was added to the cart, the issue is not specific to users of screen readers. Navigating and scrolling the Mini-Cart contents might be as complex for a sighted user as for a non-sighted user, as it depends on the input methods they use.

> I would rather tell them what and how many were added to cart

I agree, that would definitely be better! It was originally blocked by the fact that there wasn't an iAPI i18n library that would let us interpolate elements. I guess that's also why the iAPI Mini-Cart title is "Mini-Cart (items: X)" instead of "Mini-Cart (X items)" as it was before. I'm not sure what's the current status of i18n in the iAPI, but I would separate those enhancements from this one to avoid the scope of this issue growing.

---

In any case, to keep the discussion moving forward, I created a PR in https://github.com/woocommerce/woocommerce/pull/63021 that adds a `Added to cart` announcement to the Mini-Cart when it's triggered by an added-to-cart event. As mentioned, I believe it doesn't add much valuable context, but I'm happy to change my mind if you think otherwise. 😄 I would appreciate it if you could take a look and tell me if that behavior is what you had in mind or if you would implement it differently.

**Comment by samueljseay** [comment]
> Well, my point was that, as a sighted user:

> I can't see the button increment because the Mini-Cart covers the button before the new quantity is visible.

That's a good point. That seems like a flaw in the design somewhat. 🤔 

> I agree, that would definitely be better! It was originally blocked by the fact that there wasn't an iAPI i18n library that would let us interpolate elements. I guess that's also why the iAPI Mini-Cart title is "Mini-Cart (items: X)" instead of "Mini-Cart (X items)" as it was before. I'm not sure what's the current status of i18n in the iAPI, but I would separate those enhancements from this one to avoid the scope of this issue growing.

Noted, we have some options to do this, but it definitely is harder, and may not be a perfect translation for now like you say.

> In any case, to keep the discussion moving forward, I created a PR in https://github.com/woocommerce/woocommerce/pull/63021 that adds a Added to cart announcement to the Mini-Cart when it's triggered by an added-to-cart event. As mentioned, I believe it doesn't add much valuable context, but I'm happy to change my mind if you think otherwise.

Thanks for doing that, like I said I'm also not sure how critical it is to impaired users, I figure its safer to err towards keeping any additional context information. I will review :)
